File: chapter01/oneAway/OneAway.ts
function OneAway(s1: string, s2: string): boolean {
  if (Math.abs(s1.length - s2.length) > 2) return false;

  let edits = 0;
  let [i, j] = [0, 0];
  let [n, m] = [s1.length, s2.length];

  while (i < n && j < m && edits <= 1) {
    if (s1[i] !== s2[j]) {
      edits += 1;
      if (n > m) {
        i += 1
      } else if (m > n) {
        j += 1;
      } else {
        [i, j] = moveBothPointers(i, j);
      }
    } else {
      [i, j] = moveBothPointers(i, j);
    }
  }

  return edits <= 1;
}

function moveBothPointers(i: number, j: number): number[] {
  return [i + 1, j + 1];
}

console.log(OneAway('pale', 'ple'));
console.log(OneAway('pales', 'pale'));
console.log(OneAway('pale', 'bale'));
console.log(OneAway('pale', 'bake'));

File: chapter01/oneAway/oneAway.java
public class oneAway {
    public static boolean isOneWay(String s1, String s2) {
        if (s1.length() == s2.length()) {
            return isOneEditReplace(s1, s2);
        } else if (s1.length() + 1 == s2.length()) {
            return isOneEditInsert(s1, s2);
        } else if (s1.length() - 1 == s2.length()) {
            return isOneEditInsert(s2, s1);
        }

        return false;
    }

    public static boolean isOneEditReplace(String s1, String s2) {
        boolean foundDifference = false;

        for (int i = 0; i < s1.length(); i++) {
            if (s1.charAt(i) != s2.charAt(i)) {
                if (foundDifference) {
                    return false;
                }

                foundDifference = true;
            }
        }

        return true;
    }

    public static boolean isOneEditInsert(String s1, String s2) {
        int index1 = 0;
        int index2 = 0;

        while (index1 < s1.length() && index2 < s2.length()) {
            if (s1.charAt(index1) != s2.charAt(index2)) {
                if (index1 != index2) {
                    return false;
                }

            } else {
                index1++;
            }
            index2++;
        }

        return true;
    }

    public static void main(String[] args) {
        System.out.println(isOneWay("pale", "ple"));
        System.out.println(isOneWay("pales", "pale"));
        System.out.println(isOneWay("pale", "bale"));
        System.out.println(isOneWay("pale", "bake"));
    }
}
